rpgJavaInterpreter-core/src/main/kotlin/com/smeup/rpgparser/interpreter/program.kt

```diff
@@ -16,6 +16,9 @@ interface Program {
 }
 
 class RpgProgram(val cu: CompilationUnit, dbInterface: DBInterface, val name: String = "<UNNAMED RPG PROGRAM>") : Program {
+
+    private var interpreter: InternalInterpreter? = null
```

Comment:
Maybe a val with by lazy initialization would be better, avoiding !! calls

Reply:
InternalInterpreter requires a non optional SystemInterface instance.
I thought to proceed as follows:

```
class RpgProgram(val cu: CompilationUnit, dbInterface: DBInterface, val name: String = "<UNNAMED RPG PROGRAM>") : Program {

    private var systemInterface: SystemInterface? = null

    private val interpreter: InternalInterpreter by lazy {
        InternalInterpreter(this.systemInterface!!)
    }
```

and to change execute implementation like this:
```
    override fun execute(systemInterface: SystemInterface, params: LinkedHashMap<String, Value>): List<Value> {
        require(params.keys.toSet() == params().asSequence().map { it.name }.toSet()) {
            "Expected params: ${params().asSequence().map { it.name }.joinToString(", ")}"
        }
        this.systemInterface = systemInterface
```

This approach works, but I think it may be interesting for the future, to expose SystemInterface through MainExecutionContext.
In this way we are sure that SystemInterface will be immutable within its execution context

What do you think?

Reply:
I have to consider this. At the moment I think your solution is pretty good.